src/libstore/store-api.cc

```diff
@@ -822,7 +822,23 @@ std::map<StorePath, StorePath> copyPaths(ref<Store> srcStore, ref<Store> dstStor
                 MaintainCount<decltype(nrRunning)> mc(nrRunning);
                 showProgress();
                 try {
-                    copyStorePath(srcStore, dstStore, storePath, repair, checkSigs);
+                    if (dstStore->isTrusting || info->ca) {
+                        copyStorePath(srcStore, dstStore, storePath, repair, checkSigs);
+                    } else if (info->deriver && dstStore->storeDir == srcStore->storeDir) {
+                        auto drvPath = *info->deriver;
+                        auto outputMap = srcStore->queryDerivationOutputMap(drvPath);
+                        auto p = std::find_if(outputMap.begin(), outputMap.end(), [&](auto & i) {
+                            return i.second == storePath;
+                        });
+                        // drv file is always CA
+                        copyStorePath(srcStore, dstStore, drvPath, repair, checkSigs);
+                        dstStore->buildPaths({{
+                            drvPath,
+                            p != outputMap.end() ? StringSet { p->first } : StringSet {},
+                        }});
```

Comment:
It seems very unexpected to me that `copyPaths` would *build* something on the destination store...

What's the use case for this?

Reply:
admittedly this isn't the ideal place to put it, but the idea is that when we have an input-addressed path and the remote builder doesn't trust us, we ask the remote builder to build the path itself.

Reply:
@edolstra I was thinking it would be very nifty if the instead of client trying to guess what the remote builder needs and copying it over, the client could open a fresh connection *acting as the daemon* with a new sort of `RestrictedStore` so the server (acting as the client) could just directly request what it needs, as managed with regular substitution goals, etc.

It's very fancy, but this does mean we can avoid any modifications to `copyPaths`.

Reply:
This is now fixed in the last commit. Only the build hook will fall back on building things which cannot be trusted. All the other uses of `copyPaths` will never build anything, as today.